vision.md

```diff
@@ -206,26 +206,188 @@ The framework includes modules for cloud connectivity, data mapping, device
+      different locations, and instead be centralized and easy to write
+    - Configuration options are easily discoverable and documented, for example,
+      in a highly secured environment, the user is not required to access some
+      online documentation ressource, but is able to access all configuration
+      documentation via the binary shipped or derived from the thin-edge.io
+      framework
+    - Configurations are representable in multiple ways, e.g. as text or in a
+      more visual format (graph, image overview)
+1. The user os able to enable and disable functionality of the framework
+    - Not all functionalities of the framework are required in every use-case,
+      users are able to disable or even remove those parts from deployed
+      binaries
+    - For example: if only approved and audited components are allowed to be
+      used
+1. Deriving specialized implementations/binaries from the framework is possible
+    - The thin-edge.io ships a potentially large all-in-one binary with as much
+      functionality as the project provides, but
+    - Tailoring down the binary and removing unused functionality is easily
+      doable with minimal effort by a System Integrator
+    - Building a specialized binary with additional custom functionality is
+      easily doable by an Plugin Developer
+1. As little overhead as possible
+    1. in CPU time: The Framework ensures that the absolute minimum of CPU time
+       is consumed to deliver the desired functionality the user configured the
+       framework to do
+    1. in Memory usage: The Framework ensures that as little as possible memory
+       is in use, at every point in execution time, as possible
+1. Compatibility
+    - Configuration and components are written in such a way that they may be
+      changed in the future with no or minimal impact on their intended purpose
+
+### Functional requirements
+
+<!--
+The following list of functional requirements should be considered
+non-final. It might be extended in the future.
+-->
+
+1. The framework provides MQTT connectivity through a component
+1. All errors must be handled in a non-crashing way
+    - Unrecoverable errors may still cause the binary to shutdown eventually,
+      but not unexpectedly.
+1. The core implementation is written in Rust
+    - Plugins that connect to a thin-edge.io binary, may be written in another
+      programming language
+1. A deployment of thin-edge.io consists of a single binary with a single
+   configuration entry-point
+    - The final configuration may consists of more than one resource (file,
+      network resource, etc), and potentially be even loaded over the network,
+      this is left open
+1. The configuration is the single point of truth w.r.t. the initial state of
+   the components mentioned within
+    1. The default state of a plugin must be documented
+    1. A component may use a (documented) default value for a missing
+       configuration entry
```

Comment:
This is not clear to me. What is the relationship between the configuration and the initial state? If a plugin has some persisted state, I would expect the configuration giving only a path to a store containing the state of the plugin.

Reply:
So the initial state of a plugin is what the developer has decided to instantiate their component with. What this is saying, is that developers _must not_ derive this initial state from anything _but_ the configuration! So no current time or PWD or other things like that.


The other part is that if there are defaults, they need to be documented.